src/main/java/org/terasology/simpleFarming/systems/TreeAuthoritySystem.java

```diff
@@ -0,0 +1,573 @@
+		if(!rootComponent.alive || rootComponent.growthStage+1 == rootComponent.growthStages.size()) {
+			return;
+		}
+		
+		rootComponent.growthStage++;
+		if(canGenerateTree(rootEntity)) {
+			rootEntity.addOrSaveComponent(rootComponent);
+			destroyTree(rootEntity, logComponent.location);
+			generateTree(rootEntity);
+			if(rootComponent.growthStage+1 < rootComponent.growthStages.size()) {
+				TreeGrowthStage nextStage = rootComponent.growthStages.get(rootComponent.growthStage+1);
+				resetDelay(rootEntity, nextStage.minTime, nextStage.maxTime);
+			}
+		}
+	}
+	
+	/**
+	 * Reacts to an item with a CheatGrowthComponent being used on a sapling or log block, and appropriately grows or un-grows the tree.
+	 * 
+	 * @param event The event from right-clicking with an item.
+	 * @param item The item that was used.
+	 * @param cheatGrowthComponent The CheatGrowthComponent of that item.
+	 * @param itemComponent The ItemComponent of that item, not used.
+	 * 
+	 * @see #growSapling(SaplingDefinitionComponent)
+	 * @see #onRootGrowth(DelayedActionTriggeredEvent, EntityRef, LogComponent, RootComponent)
+	 */
+	@ReceiveEvent
+	public void onCheatGrowth(ActivateEvent event, EntityRef item, CheatGrowthComponent cheatGrowthComponent, ItemComponent itemComponent) {
+		EntityRef target = event.getTarget();
+		if(target.hasComponent(SaplingDefinitionComponent.class)) {
+			if(!cheatGrowthComponent.causesUnGrowth) {
+				growSapling(target.getComponent(SaplingDefinitionComponent.class));
+			}
+		} else if(target.hasComponent(LogComponent.class)) {
+			EntityRef rootEntity = target.getComponent(LogComponent.class).root;
+			RootComponent rootComponent = rootEntity.getComponent(RootComponent.class);
+			if(!rootComponent.alive) {
+				return;
+			}
+			
+			LogComponent logComponent = rootEntity.getComponent(LogComponent.class);
+			if(cheatGrowthComponent.causesUnGrowth) {
+				if(rootComponent.growthStage == 0) {
+					destroyTree(rootEntity, logComponent.location);
+					worldProvider.setBlock(logComponent.location, rootComponent.sapling);
+					EntityRef saplingEntity = blockEntityRegistry.getExistingEntityAt(logComponent.location);
+					SaplingDefinitionComponent saplingComponent = new SaplingDefinitionComponent(rootComponent, logComponent.location);
+					saplingEntity.addOrSaveComponent(saplingComponent);
+					TreeGrowthStage currentStage = saplingComponent.growthStages.get(0);
+			        resetDelay(saplingEntity, currentStage.minTime, currentStage.maxTime);
+					return;
+				}
+				rootComponent.growthStage--;
+				rootEntity.addOrSaveComponent(rootComponent);
+				destroyTree(rootEntity, logComponent.location);
+				generateTree(rootEntity);
+				TreeGrowthStage nextStage = rootComponent.growthStages.get(rootComponent.growthStage+1);
+				resetDelay(rootEntity, nextStage.minTime, nextStage.maxTime);
+			} else {
```

Comment:
See my above comment, but at the least I'd like to see both halves of this `if` in two separate methods, or containing very simple code to use other methods